runtime/transaction_executor.go

```diff
@@ -247,37 +204,52 @@ func (executor *interpreterTransactionExecutor) execute() (err error) {
 	context := executor.context
 	location := context.Location
 	codesAndPrograms := executor.codesAndPrograms
-	interpreterRuntime := executor.runtime
 
-	defer interpreterRuntime.Recover(
+	defer Recover(
 		func(internalErr Error) {
 			err = internalErr
 		},
 		location,
 		codesAndPrograms,
 	)
 
+	switch environment := environment.(type) {
+	case *interpreterEnvironment:
+		err = executor.executeWithInterpreter(environment)
+		if err != nil {
+			return newError(err, executor.context.Location, codesAndPrograms)
+		}
+		return nil
+
+	default:
+		panic(errors.NewUnexpectedError("unsupported environment: %T", environment))
+	}
```

Comment:
It seems this switch-case gets repeated in multiple places (e.g: commons for contract-function execution, script execution and transaction execution). Maybe move it to a common function?

Reply:
Not quite sure where to put it / how to abstract over it. `Executor` would need/get a `executeWithInterpreter`?